File: backend/app/services/studio_variety.py
"""Cross-generation freshness for the AI-written Studio games.

Every AI game hits the same wall: asked for the same theme twice, the model
answers with very nearly the same content. Milk, bread, eggs, apples — the
reader who buys a 60-page book meets the same shopping list on page 3 and on
page 41. The ``seed`` already in each prompt does not fix it; the model reads
the number and writes the same list anyway.

This module is the shared fix, wired into every ``studio_*`` service:

* **What was printed before is fed back in.** After a generation, the service
  remembers the labels it produced under a ``(game, user, bucket)`` key. The
  next call for that bucket lists them as already used, so the model has to
  reach further into the theme.
* **The client remembers too.** Browsers send their own ``avoid`` list, which
  covers the two things a process-local memory cannot: more than one API worker,
  and a restart. Both lists are merged here.
* **A rotating angle.** A seed-picked lens ("things tied to cold weather") moves
  the model to a different corner of the theme on the very first call, before
  any memory exists.

Services touch exactly two functions: :func:`with_variety` around the prompt
they already build, and :func:`remember` once the output is parsed.
"""

from __future__ import annotations

import hashlib
import re
from collections import OrderedDict, deque
from dataclasses import dataclass, replace
from functools import lru_cache
from typing import Any, Iterable, Mapping, Sequence

from app.services.prompt_data import (
    bullet_lines,
    int_value,
    load_config,
    rotate,
    section,
    string_list,
)

SHARED = "shared"
CONFIG_NAME = "variety"

_WHITESPACE_RE = re.compile(r"\s+")
# Trailing punctuation only: "Apples," and "apples" are the same item to a reader.
_EDGE_PUNCTUATION_RE = re.compile(r"^[^\w]+|[^\w]+$", re.UNICODE)


@lru_cache(maxsize=1)
def _config() -> Mapping[str, Any]:
    return load_config(SHARED, CONFIG_NAME)


@lru_cache(maxsize=1)
def _limits() -> Mapping[str, Any]:
    return section(_config(), "limits")


@lru_cache(maxsize=1)
def _lines() -> Mapping[str, Any]:
    return section(_config(), "lines")


@lru_cache(maxsize=1)
def _angles() -> tuple[str, ...]:
    return string_list(_config(), "angles")


def normalize_label(value: Any) -> str:
    """Collapse a label to its comparable form (empty when it carries nothing)."""
    text = _WHITESPACE_RE.sub(" ", str(value or "")).strip()
    text = _EDGE_PUNCTUATION_RE.sub("", text)
    return text[: int_value(_limits(), "maxLabelChars")].strip()


def bucket_key(*parts: Any) -> str:
    """Join the config fields that decide "same kind of content" into one key.

    Two generations only risk repeating each other when they ask for the same
    thing — same theme, same category, same era. Pass those fields; leave counts
    and layout options out, they do not change what the model writes about.
    """
    cleaned = [normalize_label(part).casefold() for part in parts]
    return "|".join(part for part in cleaned if part) or "default"


@dataclass(frozen=True)
class VarietyScope:
    """Who is generating, for which game, about what, on which attempt."""

    game: str
    user_id: str
    bucket: str
    seed: int

    def at_seed(self, seed: int) -> "VarietyScope":
        """Same memory bucket, new draw — for services that retry per batch."""
        return replace(self, seed=seed)

    @property
    def key(self) -> str:
        user = normalize_label(self.user_id).casefold() or "anonymous"
        return f"{self.game}|{user}|{self.bucket}"


class VarietyMemory:
    """Recently generated labels per scope key, newest last.

    Process-local and bounded, like :class:`RateLimiter`: it is a nudge for the
    prompt, not a source of truth. The browser's own list is what survives a
    restart or a second worker.
    """

    def __init__(self, *, per_bucket: int, max_buckets: int) -> None:
        self._per_bucket = per_bucket
        self._max_buckets = max_buckets
        self._buckets: OrderedDict[str, deque[str]] = OrderedDict()

    def remember(self, key: str, values: Iterable[str]) -> None:
        labels = [label for label in (normalize_label(v) for v in values) if label]
        if not labels:
            return
        bucket = self._buckets.get(key)
        if bucket is None:
            bucket = deque(maxlen=self._per_bucket)
            self._buckets[key] = bucket
        self._buckets.move_to_end(key)

        seen = {label.casefold() for label in bucket}
        for label in labels:
            folded = label.casefold()
            if folded in seen:
                continue
            seen.add(folded)
            # deque drops the oldest at maxlen; `seen` may keep a stale entry,
            # which only means one label is not re-added until it rotates out.
            bucket.append(label)

        while len(self._buckets) > self._max_buckets:
            self._buckets.popitem(last=False)

    def recent(self, key: str, limit: int) -> tuple[str, ...]:
        bucket = self._buckets.get(key)
        if not bucket or limit <= 0:
            return ()
        # Newest first: the most recent sheet is the one a reader would notice.
        return tuple(reversed(list(bucket)[-limit:]))

    def clear(self) -> None:
        self._buckets.clear()


_memory = VarietyMemory(
    per_bucket=int_value(_limits(), "memoryPerBucket"),
    max_buckets=int_value(_limits(), "maxBuckets"),
)


def _merge_avoid(scope: VarietyScope, client_avoid: Sequence[str], limit: int) -> list[str]:
    """Client list first (it spans sessions), then this worker's own memory."""
    merged: list[str] = []
    seen: set[str] = set()
    for raw in list(client_avoid) + list(_memory.recent(scope.key, limit)):
        label = normalize_label(raw)
        folded = label.casefold()
        if not label or folded in seen:
            continue
        seen.add(folded)
        merged.append(label)
        if len(merged) >= limit:
            break
    return merged


def variety_angle(seed: int) -> str:
    """The lens this seed asks the model to look through."""
    return rotate(_angles(), seed)


def variety_nonce(scope: VarietyScope) -> str:
    """Short token that differs per scope+seed, so two calls never look alike."""
    digest = hashlib.sha1(f"{scope.key}:{scope.seed}".encode("utf-8")).hexdigest()
    return digest[:6]


def variety_lines(
    scope: VarietyScope,
    *,
    client_avoid: Sequence[str] = (),
    limit: int | None = None,
) -> str:
    """Render the freshness block to drop into a prompt template.

    Always returns the angle and the variation token; the "already used" list
    appears only once there is history for this scope.
    """
    lines = _lines()
    max_avoid = limit if limit is not None else int_value(_limits(), "promptAvoidCount")
    avoided = _merge_avoid(scope, client_avoid, max_avoid)

    body = [
        str(lines["angle"]).format(angle=variety_angle(scope.seed)),
        str(lines["nonce"]).format(nonce=variety_nonce(scope)),
    ]
    if avoided:
        body.insert(0, str(lines["avoid"]).format(avoided=", ".join(avoided)))
        body.insert(1, str(lines["avoidTail"]))

    return f"{lines['header']}\n{bullet_lines(body)}"


def with_variety(
    prompt: str,
    scope: VarietyScope,
    *,
    client_avoid: Sequence[str] = (),
    limit: int | None = None,
) -> str:
    """Append the freshness block to a finished prompt.

    Appending rather than threading a placeholder through every prompt template:
    each service keeps its own wording untouched, and the block lands last,
    where the model reads it right before answering.
    """
    block = variety_lines(scope, client_avoid=client_avoid, limit=limit)
    return f"{prompt.rstrip()}\n\n{block}\n"


def remember(scope: VarietyScope, values: Iterable[Any]) -> None:
    """Record what this generation produced so the next one has to differ."""
    _memory.remember(scope.key, (str(v) for v in values))


def recent(scope: VarietyScope, limit: int | None = None) -> tuple[str, ...]:
    max_avoid = limit if limit is not None else int_value(_limits(), "promptAvoidCount")
    return _memory.recent(scope.key, max_avoid)


def reset_memory() -> None:
    """Test helper — drops every remembered bucket."""
    _memory.clear()

File: frontend/src/utils/studio/studio-variety.ts
/**
 * What the AI templates have already printed, kept per template + theme.
 *
 * The seed in a generate request buys nothing from a language model: ask for
 * "a picnic shopping list" twice and the same eight items come back, so a book
 * repeats itself every few pages. Every AI prefetch therefore sends what it
 * printed last time in `avoid`, and the service turns that into a
 * "do not repeat these" line in the prompt.
 *
 * This lives in the browser rather than only on the server because a book is
 * built over several sittings, against however many API workers happen to
 * answer. localStorage is the only place that sees all of it.
 *
 * Storage is best-effort by design: a blocked or full store degrades to an
 * in-memory map for the session, never to a failed generation.
 */

const STORAGE_KEY = 'studio:recent-content:v1'

/** Labels sent to the server per request. Matches the backend's prompt cap. */
export const STUDIO_AVOID_LIMIT = 60
/** Labels kept per template + theme. A few sheets' worth of history. */
const MEMORY_PER_KEY = 200
/** Distinct template + theme buckets kept before the oldest is dropped. */
const MAX_KEYS = 200
const MAX_LABEL_CHARS = 60

/** Newest last, so a bucket reads as a printing history. */
type RecentStore = Record<string, string[]>

/**
 * Session fallback, used once localStorage proves unavailable or unwritable
 * (private mode, quota, SSR, tests). Non-null means "read and write here".
 */
let fallbackStore: RecentStore | null = null

function normalizeLabel(value: unknown): string {
  return String(value ?? '')
    .replace(/\s+/g, ' ')
    .trim()
    .replace(/^[^\p{L}\p{N}]+|[^\p{L}\p{N}]+$/gu, '')
    .slice(0, MAX_LABEL_CHARS)
    .trim()
}

function storage(): Storage | null {
  try {
    return typeof localStorage === 'undefined' ? null : localStorage
  } catch {
    return null
  }
}

function readStore(): RecentStore {
  if (fallbackStore) return fallbackStore
  const store = storage()
  if (!store) {
    fallbackStore = {}
    return fallbackStore
  }
  try {
    const raw = store.getItem(STORAGE_KEY)
    if (!raw) return {}
    const parsed = JSON.parse(raw) as unknown
    if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) return {}
    const out: RecentStore = {}
    for (const [key, value] of Object.entries(parsed as Record<string, unknown>)) {
      if (!Array.isArray(value)) continue
      out[key] = value.filter((entry): entry is string => typeof entry === 'string')
    }
    return out
  } catch {
    return {}
  }
}

function writeStore(next: RecentStore): void {
  const store = fallbackStore ? null : storage()
  if (!store) {
    fallbackStore = next
    return
  }
  try {
    store.setItem(STORAGE_KEY, JSON.stringify(next))
  } catch {
    // Quota or a locked-down browser: keep the history for this session only.
    fallbackStore = next
  }
}

/**
 * Bucket the config fields that decide *what* gets written — theme, category,
 * era. Counts and layout options belong out of the key: they change the sheet,
 * not the words on it.
 */
export function studioVarietyKey(templateKey: string, ...parts: unknown[]): string {
  const tail = parts
    .map((part) => normalizeLabel(part).toLowerCase())
    .filter(Boolean)
    .join('|')
  return tail ? `${templateKey}|${tail}` : `${templateKey}|default`
}

/** Recently printed labels for this key, newest first. */
export function studioAvoidList(key: string, limit: number = STUDIO_AVOID_LIMIT): string[] {
  if (limit <= 0) return []
  const bucket = readStore()[key]
  if (!bucket?.length) return []
  return bucket.slice(-limit).reverse()
}

/** Record what a generation produced, so the next one is asked for something else. */
export function rememberStudioContent(key: string, values: Iterable<unknown>): void {
  const labels: string[] = []
  for (const value of values) {
    const label = normalizeLabel(value)
    if (label) labels.push(label)
  }
  if (labels.length === 0) return

  const store = readStore()
  const existing = store[key] ?? []
  const seen = new Set(existing.map((label) => label.toLowerCase()))
  const merged = [...existing]
  for (const label of labels) {
    const folded = label.toLowerCase()
    if (seen.has(folded)) continue
    seen.add(folded)
    merged.push(label)
  }

  const next: RecentStore = { ...store, [key]: merged.slice(-MEMORY_PER_KEY) }

  // Object key order is insertion order, so the oldest bucket is the first one.
  const keys = Object.keys(next)
  if (keys.length > MAX_KEYS) {
    for (const stale of keys.slice(0, keys.length - MAX_KEYS)) {
      if (stale !== key) delete next[stale]
    }
  }

  writeStore(next)
}

/** Test helper — drops every remembered bucket. */
export function clearStudioRecentContent(): void {
  fallbackStore = null
  const store = storage()
  try {
    store?.removeItem(STORAGE_KEY)
  } catch {
    // Nothing to clear if the store refuses us.
  }
}
